### ISSUE TITLE
[BUG] FlutterToastPage context is used after disposed

### ISSUE THREAD

**Issue by narumishi** [jhomlala/catcher#175]
First of all, please add all platforms in `getSupportedPlatforms`.

**Describe the bug**
It may happen when several errors raised in a very short time. 

```dart
void testFunction() async {
  Catcher.reportCheckedError('Test Error 1', StackTrace.current);
  await Future.delayed(Duration(milliseconds: 50));
  Catcher.reportCheckedError('Test Error 2', StackTrace.current);
  await Future.delayed(Duration(milliseconds: 50));
  Catcher.reportCheckedError('Test Error 3', StackTrace.current);
  await Future.delayed(Duration(milliseconds: 50));
  Catcher.reportCheckedError('Test Error 4', StackTrace.current);
  await Future.delayed(Duration(milliseconds: 50));
  return;
}
```

Error 1:
```
flutter: [2021-05-03 00:13:59.111809 | ConsoleHandler | INFO] ---------- ERROR ----------
flutter: [2021-05-03 00:13:59.111945 | ConsoleHandler | INFO] This widget has been unmounted, so the State no longer has a context (and should be considered defunct).
Consider canceling any active work during "dispose" or using the "mounted" getter to determine if the State is still active.
flutter: [2021-05-03 00:13:59.111998 | ConsoleHandler | INFO] 
flutter: [2021-05-03 00:13:59.112026 | ConsoleHandler | INFO] ------- STACK TRACE -------
flutter: [2021-05-03 00:13:59.112156 | ConsoleHandler | INFO] #0      State.context.<anonymous closure> (package:flutter/src/widgets/framework.dart:909:9)
flutter: [2021-05-03 00:13:59.112197 | ConsoleHandler | INFO] #1      State.context (package:flutter/src/widgets/framework.dart:915:6)
flutter: [2021-05-03 00:13:59.112223 | ConsoleHandler | INFO] #2      _FlutterToastPageState.showToast.<anonymous closure> (package:catcher/handlers/toast_handler.dart:160:22)
flutter: [2021-05-03 00:13:59.112259 | ConsoleHandler | INFO] #3      new Future.delayed.<anonymous closure> (dart:async/future.dart:315:39)
flutter: [2021-05-03 00:13:59.112292 | ConsoleHandler | INFO] #4      _rootRun (dart:async/zone.dart:1420:47)
flutter: [2021-05-03 00:13:59.112317 | ConsoleHandler | INFO] #5      _CustomZone.run (dart:async/zone.dart:1328:19)
flutter: [2021-05-03 00:13:59.112344 | ConsoleHandler | INFO] #6      _CustomZone.runGuarded (dart:async/zone.dart:1236:7)
flutter: [2021-05-03 00:13:59.112368 | ConsoleHandler | INFO] #7      _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1276:23)
flutter: [2021-05-03 00:13:59.112394 | ConsoleHandler | INFO] #8      _rootRun (dart:async/zone.dart:1428:13)
flutter: [2021-05-03 00:13:59.112419 | ConsoleHandler | INFO] #9      _CustomZone.run (dart:async/zone.dart:1328:19)
flutter: [2021-05-03 00:13:59.112445 | ConsoleHandler | INFO] #10     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:1260:23)
flutter: [2021-05-03 00:13:59.112470 | ConsoleHandler | INFO] #11     Timer._createTimer.<anonymous closure> (dart:async-patch/timer_patch.dart:18:15)
flutter: [2021-05-03 00:13:59.112499 | ConsoleHandler | INFO] #12     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:395:19)
flutter: [2021-05-03 00:13:59.112532 | ConsoleHandler | INFO] #13     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:426:5)
flutter: [2021-05-03 00:13:59.112566 | ConsoleHandler | INFO] #14     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
flutter: [2021-05-03 00:13:59.112598 | ConsoleHandler | INFO] 
flutter: [2021-05-03 00:13:59.112639 | ConsoleHandler | INFO] ======================================================================
```

Or maybe Error 2
```
flutter: [2021-05-03 00:42:02.729936 | ConsoleHandler | INFO] ---------- ERROR ----------
flutter: [2021-05-03 00:42:02.730296 | ConsoleHandler | INFO] Looking up a deactivated widget's ancestor is unsafe.
At this point the state of the widget's element tree is no longer stable.
To safely refer to a widget's ancestor in its dispose() method, save a reference to the ancestor by calling dependOnInheritedWidgetOfExactType() in the widget's didChangeDependencies() method.
flutter: [2021-05-03 00:42:02.730455 | ConsoleHandler | INFO] 
flutter: [2021-05-03 00:42:02.730553 | ConsoleHandler | INFO] ------- STACK TRACE -------
flutter: [2021-05-03 00:42:02.730785 | ConsoleHandler | INFO] #0      Element._debugCheckStateIsActiveForAncestorLookup.<anonymous closure> (package:flutter/src/widgets/framework.dart:3966:9)
flutter: [2021-05-03 00:42:02.730844 | ConsoleHandler | INFO] #1      Element._debugCheckStateIsActiveForAncestorLookup (package:flutter/src/widgets/framework.dart:3980:6)
flutter: [2021-05-03 00:42:02.730889 | ConsoleHandler | INFO] #2      Element.findAncestorStateOfType (package:flutter/src/widgets/framework.dart:4027:12)
flutter: [2021-05-03 00:42:02.730927 | ConsoleHandler | INFO] #3      Overlay.of (package:flutter/src/widgets/overlay.dart:294:19)
flutter: [2021-05-03 00:42:02.731064 | ConsoleHandler | INFO] #4      FToast._showOverlay (package:fluttertoast/fluttertoast.dart:145:13)
flutter: [2021-05-03 00:42:02.731227 | ConsoleHandler | INFO] #5      FToast.removeCustomToast (package:fluttertoast/fluttertoast.dart:160:5)
flutter: [2021-05-03 00:42:02.731438 | ConsoleHandler | INFO] #6      FToast._showOverlay.<anonymous closure>.<anonymous closure> (package:fluttertoast/fluttertoast.dart:149:9)
flutter: [2021-05-03 00:42:02.731572 | ConsoleHandler | INFO] #7      new Future.delayed.<anonymous closure> (dart:async/future.dart:315:39)
flutter: [2021-05-03 00:42:02.731676 | ConsoleHandler | INFO] #8      _rootRun (dart:async/zone.dart:1420:47)
flutter: [2021-05-03 00:42:02.731752 | ConsoleHandler | INFO] #9      _CustomZone.run (dart:async/zone.dart:1328:19)
flutter: [2021-05-03 00:42:02.731828 | ConsoleHandler | INFO] #10     _CustomZone.runGuarded (dart:async/zone.dart:1236:7)
flutter: [2021-05-03 00:42:02.732035 | ConsoleHandler | INFO] #11     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1276:23)
flutter: [2021-05-03 00:42:02.732131 | ConsoleHandler | INFO] #12     _rootRun (dart:async/zone.dart:1428:13)
flutter: [2021-05-03 00:42:02.732182 | ConsoleHandler | INFO] #13     _CustomZone.run (dart:async/zone.dart:1328:19)
flutter: [2021-05-03 00:42:02.732261 | ConsoleHandler | INFO] #14     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:1260:23)
flutter: [2021-05-03 00:42:02.732302 | ConsoleHandler | INFO] #15     Timer._createTimer.<anonymous closure> (dart:async-patch/timer_patch.dart:18:15)
flutter: [2021-05-03 00:42:02.732343 | ConsoleHandler | INFO] #16     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:395:19)
flutter: [2021-05-03 00:42:02.732383 | ConsoleHandler | INFO] #17     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:426:5)
flutter: [2021-05-03 00:42:02.732427 | ConsoleHandler | INFO] #18     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
flutter: [2021-05-03 00:42:02.732464 | ConsoleHandler | INFO] 
flutter: [2021-05-03 00:42:02.732504 | ConsoleHandler | INFO] ======================================================================
```

I think the `FlutterToastPage` is unnecessary. This widget here is to pass `context` to `FToast` right? It seems not safe here. Maybe it will be popped by other code. So its context is never valid: `Navigator.of(context)` and `FToast`. 

Why not directly pass context in `ToastHandler.handle`?

See the changed code, at least the test code above passed.
```dart
import 'package:catcher/core/application_profile_manager.dart';
import 'package:catcher/model/platform_type.dart';
import 'package:catcher/model/report.dart';
import 'package:catcher/model/report_handler.dart';
import 'package:catcher/model/toast_handler_gravity.dart';
import 'package:catcher/model/toast_handler_length.dart';
import 'package:flutter/material.dart';
import 'package:fluttertoast/fluttertoast.dart';

class ToastHandler extends ReportHandler {
  final ToastHandlerGravity gravity;
  final ToastHandlerLength length;
  final Color backgroundColor;
  final Color textColor;
  final double textSize;
  final String? customMessage;

  ToastHandler({
    this.gravity = ToastHandlerGravity.bottom,
    this.length = ToastHandlerLength.long,
    this.backgroundColor = Colors.black87,
    this.textColor = Colors.white,
    this.textSize = 12,
    this.customMessage,
  });

  FToast fToast = FToast();

  @override
  Future<bool> handle(Report error, BuildContext? buildContext) async {
    if (ApplicationProfileManager.isAndroid() ||
        ApplicationProfileManager.isIos() ||
        ApplicationProfileManager.isWeb()) {
      Fluttertoast.showToast(
          msg: _getErrorMessage(error),
          toastLength: _getLength(),
          gravity: _getGravity(),
          timeInSecForIosWeb: _getLengthIos(),
          backgroundColor: backgroundColor,
          textColor: textColor,
          fontSize: textSize);
    } else {
      // If error raised before app initiated
      if (buildContext == null) {
        return false;
      }
      fToast.init(buildContext);
      Duration duration = Duration(seconds: _getLengthIos());
      Future.delayed(const Duration(milliseconds: 500), () {
        fToast.showToast(
          child: Container(
            color: backgroundColor,
            child: Text(
              _getErrorMessage(error),
              style: TextStyle(
                color: textColor,
                fontSize: textSize,
              ),
            ),
          ),
          gravity: _getGravity(),
          toastDuration: duration,
        );
      });
    }

    return true;
  }

  ToastGravity _getGravity() {
    switch (gravity) {
      case ToastHandlerGravity.bottom:
        return ToastGravity.BOTTOM;
      case ToastHandlerGravity.center:
        return ToastGravity.CENTER;
      case ToastHandlerGravity.top:
        return ToastGravity.TOP;
    }
  }

  Toast _getLength() {
    if (length == ToastHandlerLength.long) {
      return Toast.LENGTH_LONG;
    } else {
      return Toast.LENGTH_SHORT;
    }
  }

  int _getLengthIos() {
    if (length == ToastHandlerLength.long) {
      return 5;
    } else {
      return 1;
    }
  }

  String _getErrorMessage(Report error) {
    if (customMessage?.isNotEmpty == true) {
      return customMessage!;
    } else {
      return "${localizationOptions.toastHandlerDescription} ${error.error}";
    }
  }

  @override
  List<PlatformType> getSupportedPlatforms() => [
        PlatformType.web,
        PlatformType.android,
        PlatformType.iOS,
        PlatformType.linux,
        PlatformType.macOS,
        PlatformType.windows,
      ];

  @override
  bool isContextRequired() {
    return true;
  }
}
```
**Catcher version**
- Version: 0.6.5, macOS

Another thing said in the closed issue: Add error detail in ErrorWidget too
https://github.com/jhomlala/catcher/issues/163#issuecomment-822032376

**Comment by jhomlala** [comment]
@narumishi FlutterToastPage is required because of the overlay in FToast:

![image](https://user-images.githubusercontent.com/1329033/116901128-d9ac6780-ac39-11eb-99d7-ea5d4c2f6107.png)

As you can see, context passed in `fToast.init(context);` is used to access Overlay. So when we're using navigation and page route then we're adding overlay in widget tree which will be used by FToast.

**Comment by narumishi** [comment]
Yes, I do know context is required, but the context of FlutterToastPageState you passed may be used after it is disposed.

You can see the code I offer above, use the buildContext directly, since it is retrieved from navigator, it should always be valid. And it works.

**Comment by jhomlala** [comment]
@narumishi When I try to use your code, I see this error:
```dart
I/flutter (10482): [2021-05-03 18:16:39.530857 | ConsoleHandler | INFO] ---------- ERROR ----------
I/flutter (10482): [2021-05-03 18:16:39.530956 | ConsoleHandler | INFO] Null check operator used on a null value
I/flutter (10482): [2021-05-03 18:16:39.531050 | ConsoleHandler | INFO] 
I/flutter (10482): [2021-05-03 18:16:39.531143 | ConsoleHandler | INFO] ------- STACK TRACE -------
I/flutter (10482): [2021-05-03 18:16:39.531426 | ConsoleHandler | INFO] #0      FToast._showOverlay (package:fluttertoast/fluttertoast.dart:145:25)
I/flutter (10482): [2021-05-03 18:16:39.531535 | ConsoleHandler | INFO] #1      FToast.showToast (package:fluttertoast/fluttertoast.dart:198:25)
I/flutter (10482): [2021-05-03 18:16:39.531676 | ConsoleHandler | INFO] #2      ToastHandler.handle.<anonymous closure> (package:catcher/handlers/toast_handler.dart:49:16)
I/flutter (10482): [2021-05-03 18:16:39.531791 | ConsoleHandler | INFO] #3      new Future.delayed.<anonymous closure> (dart:async/future.dart:315:39)
I/flutter (10482): [2021-05-03 18:16:39.531946 | ConsoleHandler | INFO] #4      _rootRun (dart:async/zone.dart:1346:47)
I/flutter (10482): [2021-05-03 18:16:39.532048 | ConsoleHandler | INFO] #5      _CustomZone.run (dart:async/zone.dart:1258:19)
I/flutter (10482): [2021-05-03 18:16:39.532136 | ConsoleHandler | INFO] #6      _CustomZone.runGuarded (dart:async/zone.dart:1162:7)
I/flutter (10482): [2021-05-03 18:16:39.532233 | ConsoleHandler | INFO] #7      _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1202:23)
I/flutter (10482): [2021-05-03 18:16:39.532328 | ConsoleHandler | INFO] #8      _rootRun (dart:async/zone.dart:1354:13)
I/flutter (10482): [2021-05-03 18:16:39.532424 | ConsoleHandler | INFO] #9      _CustomZone.run (dart:async/zone.dart:1258:19)
I/flutter (10482): [2021-05-03 18:16:39.532519 | ConsoleHandler | INFO] #10     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:1186:23)
I/flutter (10482): [2021-05-03 18:16:39.532609 | ConsoleHandler | INFO] #11     Timer._createTimer.<anonymous closure> (dart:async-patch/timer_patch.dart:18:15)
I/flutter (10482): [2021-05-03 18:16:39.532737 | ConsoleHandler | INFO] #12     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:395:19)
I/flutter (10482): [2021-05-03 18:16:39.532833 | ConsoleHandler | INFO] #13     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:426:5)
I/flutter (10482): [2021-05-03 18:16:39.532917 | ConsoleHandler | INFO] #14     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
I/flutter (10482): [2021-05-03 18:16:39.533021 | ConsoleHandler | INFO] 
I/flutter (10482): [2021-05-03 18:16:39.533120 | ConsoleHandler | INFO] ======================================================================
```

fluttertoast.dart:145 is ` Overlay.of(context!)!.insert(_entry!);`

**Comment by narumishi** [comment]
I will check again after back home.

If use original code, The example testFunction will raise an extra error right?

**Comment by jhomlala** [comment]
Yes, I was able te reproduce this issue with your `testFunction`.

**Comment by narumishi** [comment]
I was testing on my own project, where my `MaterialApp's builder wraps child in `FlutterEasyLoading` (with `Overlay` child widget). So `Overlay.of(context)` will always find an overlay in widget tree.

But in the simplest example, the  `context = navigatorKey?.currentState?.overlay?.context`. There is no Overlay **ancestor**, but itself is the context of that overlay.

**Comment by jhomlala** [comment]
Fixed in Catcher 0.6.6.

**Comment by narumishi** [comment]
The solution has something unexpected:
- some toasts may be skipped
- The error `Looking up a deactivated widget's ancestor is unsafe` is still unsolved
